keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,510 @@
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+We will build an extensible system which controls network traffic from the Kube API Server.
+We will add a network proxy system. The KAS can be configured to send traffic
+(or not) to one or more of the proxies. Users can drop in custom proxies if the
+default behavior is insufficient.
+
+## Motivation
+
+Kubernetes has outgrown the [SSH tunnels](https://github.com/kubernetes/kubernetes/issues/54076).
+They complicate KAS code and only one cloud provider implemented them.
+After a year of deprecation time, they will be removed in an upcoming release.
+
+In retrospect, having an explicit level of indirection that separates user-initiated network traffic from API
+server-initiated traffic is a useful concept.
+Cloud providers want to control how API server to pod, node and service network traffic is implemented.
+Cloud providers may choose to run their API server (control network) and the cluster nodes (cluster network)
+on isolated networks. The control and cluster networks may have overlapping IP addresses.
+There for they require a non IP routing layer (SSH tunnel are an example).
+Adding this layer enables metadata audit logging. It allows validation of outgoing API server connections.
+Structuring the API server in this way is a forcing function for keeping architectural layering violations out of apiserver.
+In combination with a firewall, this separation of networks protects against security concerns such as
+[Security Impact of Kubernetes API server external IP address proxying](https://groups.google.com/d/msg/kubernetes-security-announce/tyd-MVR-tY4/tyREP9-qAwAJ).
+
+### Goals
+
+Delete the SSH Tunnel/Node Dialer code from Kube APIServer.
+Enable admins to fix https://groups.google.com/d/msg/kubernetes-security-announce/tyd-MVR-tY4/tyREP9-qAwAJ.
+Allow isolation of the Control network from the Cluster network.
```

Comment:
I don't want to pile on here, but could this also allow the deprecation of https://github.com/kubernetes/kubernetes/blob/d1031b18fc83146dd8418c17df289d09d10d29e5/staging/src/k8s.io/apimachinery/pkg/util/net/http.go#L273-L309

Or is this used by more than the API server?

Reply:
Agree that this is a further benefit for doing this work.